### ISSUE TITLE
Feature Request: Export to File

### ISSUE THREAD

**Issue by ScanuNicco** [Akylas/com.akylas.documentscanner#1]
In the current version of the app, pressing the PDF button opens the newly exported PDF in the default PDF viewer, and mine doesn't have a save to disk button because I'm using a non-gapps rom. I would love to see an option to export the PDF directly to a file rather than opening it.

Thanks for making this app by the way. There aren't very many open-source scanners that are modern. The app looks great so far!

**Comment by IzzySoft** [comment]
> didnt know the app was on izydroid yet :)

It isn't. But give me a proper app description, and it can be :wink:

**Comment by farfromrefug** [comment]
@IzzySoft yes i saw that after . But i knew you would come by  :D How did you find that issue? you have detector or something?
 I have already filled fastlane metadata:
 * https://github.com/Akylas/com.akylas.documentscanner/blob/master/fastlane/metadata/android/en-US/full_description.txt
 * https://github.com/Akylas/com.akylas.documentscanner/blob/master/fastlane/metadata/android/en-US/short_description.txt
 Isnt that what you read for showing info on the repo?

**Comment by IzzySoft** [comment]
> How did you find that issue? you have detector or something?

"Something", yes. One of my agents told me to have a look :wink:

> I have already filled fastlane metadata:

Great! Just your `full_description.txt` will become a "wall of text" (when processed as Markdown) or not look "that great" (when taken as-is). May I propose a slight adjustment( which I'll for now apply here)?

```html
<p>The <i>OSS Document Scanner</i> is an Open Source app to scan all your documents. You either scan using your camera or by importing an image. The app will automatically detect you document within the photo and will crop the image.</p>
<p>Once the document is created you can detect text within the document using OCR.</p>
<p>You can also share your document as a PDF. If you want you can synchronize the app data with a webdav server (like nextloud) to never loose anything!</p>
```

For German (taken from what Google shipped, slightly adjusted):

`short_description.txt`: Scannen Sie all Ihre Dokumente

`full_description.txt`:

```html
<p>Der <i>OSS Document Scanner</i> ist eine Open-Source-App zum Scannen all Ihrer Dokumente. Sie scannen entweder mit Ihrer Kamera oder importieren ein Bild. Die App erkennt Ihr Dokument automatisch im Foto und schneidet das Bild zu.</p>
<p>Sobald das Dokument erstellt ist, können Sie mithilfe von OCR Text im Dokument erkennen lassen.</p>
<p>Sie können Ihr Dokument auch als PDF teilen. Wenn Sie möchten, können Sie die App-Daten mit einem WebDAV-Server (wie Nextloud) synchronisieren, um nie etwas zu verlieren!</p>
```

Giving your tags "proper names" (e.g. `v<versionName>` or `v<versionCode>` or `c<versionCode`) might be a good idea as well.

Here's what will go live around 7 pm UTC in my repo:

![image](https://github.com/Akylas/com.akylas.documentscanner/assets/6781438/359e02b4-525b-4127-a533-5564bb0d1799)

And you can [ignore this one](https://www.virustotal.com/gui/file/fc027a75904a7d5d6cf909d62227fe73539d4d84bed76632cfacaf628a48f8b2/detection/f-fc027a75904a7d5d6cf909d62227fe73539d4d84bed76632cfacaf628a48f8b2-1701788187) – false positive IMHO, as that engine has from time to time…

**Comment by farfromrefug** [comment]
@IzzySoft when you are saying to give proper tags name you are talking about this ?
![Screenshot from 2023-12-05 20-50-39](https://github.com/Akylas/com.akylas.documentscanner/assets/655344/f871103f-d4d9-49d2-81e1-bd80859af035)
Cause i need that syntax. I use it to auto generate changelogs. It is a way to find the commits from last version (ios or android, alpha/beta ... ) to the current one and generate changelog from the commits.
Is it an issue?

**Comment by IzzySoft** [comment]
> will update my metadata thanks to what you said 

Great! Let me know when it's in place, so I can tell my updater.

> do you support multiple apps from the same repo.

Only if it cannot be avoided. It can lead to several complications. For example with the metadata, as you already found out. Or with tag names (as not all apps will have all their releases together, and then versions drift apart too). Such things can be solved, though, at the price of more complex setups. Like, you can use prefixes to tag names, so e.g. tag `app1_1.2.3` would indicate version 1.2.3 of app1, while a day later `app2_0.3.5` could be released in the same manner – and my updater could keep them apart.

For metadata: My repo supports custom locations for fastlane (but no symlinks inside), F-Droid's does not. My [Fastlane Cheat Sheet](https://gitlab.com/-/snippets/1895688) has some clues how to solve that. Depending on your setup:

* if you use build flavors to produce the different apps, you could use `/src/<buildFlavor>/fastlane/metadata/android/<locale>`
* otherwise, `/metadata/<packageName>/<locale>` would be a good fit

Not sure why you would need to put multiple apps into the very same git repo, though – always better for each to have its own. If there's a big part of common code, that's what git submodules are intended for.

> Cause i need that syntax.

I can give no guarantees that my updater will handle those, or handle those always correctly. Especially as you use multiple levels there (I know about "feature tags" like `featureName/tag` – but I have never seen them nested the way you do. My updater usually goes by the tag name to see if that indicates a version newer than already in my repo. I can tell it to skip that check, but that would then also make it impossible to have multiple different apps there the way I outlined above. That could be partly worked around by including the app names (or other specific patterns) in the APK file names – but with the way Github sometimes deals with attachments, no guarantees that there won't be releases ignored as my updater thinks they are not newer that what's already there locally.

**TL;DR:** Keep things simple wherever you can, especially in this context. The more complex you make them, the more chances that things can go wrong.

**Comment by ScanuNicco** [comment]
> @ScanuNicco thanks a lot for your comment ! didnt know the app was on izydroid yet :) as for the PDF there is an hidden to save to download (I am using non gapps too). you can long press that same button and it will be saved to download.
> 
> if you have things you would like added dont hesitate to create issues.

Alight, I'll give that a shot. The only other question I had is what does the app require Phone permissions for?

**Comment by farfromrefug** [comment]
@ScanuNicco it should not. will check might come from  a dependency. did it ask for permission at runtime or are you talking about manifest permission reported by fdroid or other app ?

**Comment by farfromrefug** [comment]
@IzzySoft  ok and you define the configuration on a per project/repo basis? Can it be set through a file i set up on my repo ? (so that you dont need to do it).
As for the libs they come from libraries i build myself:
* **/com/facebook/samples/zoomable**:  https://github.com/nativescript-community/ui-image/tree/master/packages/zoomimage/platforms/android/java/com/facebook/samples/zoomable
* **/com/nativescript/cameraview**:  https://github.com/nativescript-community/ui-cameraview/tree/main/packages/ui-cameraview/platforms/android/java/com/nativescript/cameraview
* 
* **/com/nativescript/md5**:  https://github.com/nativescript-community/md5/tree/main/packages/md5/platforms/android/java/com/nativescript/md5
* 
* **/com/nativescript/zoomimage**:  there should be nothing in this one but it also comes from https://github.com/nativescript-community/ui-image/tree/master/packages/zoomimage

Is there anyway i can make it so you find those location automatically? (like putting the info somewhere while building my libs?)

**Comment by user8446** [comment]
Hi @farfromrefug that you for developing this!

In my workflow I too need to export to different directories so being able to directly do that directly without opening a PDF viewer first would be great!

You mentioned long pressing a button to do this... which one? I long pressed the PDF button in the app but it didn't do anything.

**Comment by farfromrefug** [comment]
@user8446 indeed there is a bug and the long press is not working. I will fix so that it saves the pdf in the download folder. Now can you explain your workflow? Maybe i can make it easier.
I am still trying to figure out the best approach to allow:
* fast pdf extract opening for share
* a way to save that pdf on disk
* may be a way to customize the output location
Wondering how to do that best UX wise

**Comment by user8446** [comment]
Hi, basically a Save To button from this screen:

![Screenshot_20231207-114131](https://github.com/Akylas/com.akylas.documentscanner/assets/18019245/83e49c8f-edce-4cdb-8669-3a147410ae3d)

From there, the use gets the file manager so the user can pick the directory on where to save the PDF on the phone.  Personally, I save to multiple different directories.

**Comment by farfromrefug** [comment]
@user8446 what s your phone ? the layout on your phone is not good

**Comment by user8446** [comment]
Pixel 6a on Android 14.

I did increase the "display size" by one notch and turn on "bold text" so maybe it's the scaling?

![Screenshot_20231207-124922](https://github.com/Akylas/com.akylas.documentscanner/assets/18019245/1d0ac6aa-e7c2-4219-8ec2-e0cdaf9fb1f9)

**Comment by ScanuNicco** [comment]
> @user8446 indeed there is a bug and the long press is not working. I will fix so that it saves the pdf in the download folder. Now can you explain your workflow? Maybe i can make it easier. I am still trying to figure out the best approach to allow:
> 
>     * fast pdf extract opening for share
> 
>     * a way to save that pdf on disk
> 
>     * may be a way to customize the output location
>       Wondering how to do that best UX wise

I would be happy with a default save location in settings as long as there was a way to name the file I am saving.

**Comment by user8446** [comment]
Sure, from the screen above a Save To which would open Files on the phone (I'm guessing sending the intent would just open whatever file manager the user has for their default?). From there the user can pick where to save.  A default filename would be sent that the user can override.

Personally, I save in the structure like so:

```
Directory1
Directory2
Directory3
Directory4
  |
  Folder1
  Folder2
  Folder3
  etc.
Directory5
  |
  Folder1
  Folder2
  Folder3
  etc.
```

I simply navigate to where the scan needs to save. Folders then sync back to PC.

**Comment by ScanuNicco** [comment]
I just tried out the new build and this feature seems to be working quite well now. I'm going to close this issue now.

**Comment by user8446** [comment]
I just tested and yes the export button now opens the export dialog.

However, when changing folders I have to go through and accept the permission for the folder every time. It should persist after accepting unless it's a new folder that hasn't been saved to before.

**Comment by farfromrefug** [comment]
@user8446 if you click on the export folder to select one the  it opens a dialog I dont have control over.

**Comment by farfromrefug** [comment]
@IzzySoft i have users (even though i only few users right now) already notifying me about the trojan warning from [here](https://www.virustotal.com/gui/file/fc027a75904a7d5d6cf909d62227fe73539d4d84bed76632cfacaf628a48f8b2/detection/f-fc027a75904a7d5d6cf909d62227fe73539d4d84bed76632cfacaf628a48f8b2-1701788187). Is there anything i can do to make sure i dont get that warning?

**Comment by user8446** [comment]
I will open a separate issue so others can help as well....

**Comment by IzzySoft** [comment]
Not sure what you can do about that. To me, it's clearly a false positive. VT has 60+ engines, and as long as less than 5 report something, it's very much in doubt. There are a few of those engines which are notorious with such false alerts – this example shows one of them. From the big brands, Avast has it often, too. MaxSecure and SangFor I've encountered often. Others rarely.

I don't know how communicative those vendors are. or how much sense it makes contacting them to report the false positive so they can update their signatures. Sometimes triggering a rescan a (few) day(s) later even makes the mark disappear – other times it makes one of the other 3..4 adding their false flags as well…

**Comment by farfromrefug** [comment]
@IzzySoft thanks for taking the time to answer. guess I ll live with it ;)

**Comment by IzzySoft** [comment]
Probably the most practicable solution, yeah. Next release (or next (re)scan) might already have a clean result again.
